sway/desktop/output.c

```diff
@@ -427,7 +540,29 @@ static void render_container_simple(struct sway_output *output,
  */
 static void render_container_tabbed(struct sway_output *output,
 		struct sway_container *con) {
-	// TODO
+	struct sway_seat *seat = input_manager_current_seat(input_manager);
+	struct sway_container *focus = seat_get_focus(seat);
+	struct wlr_texture **title_texture = get_title_textures(con);
+
+	for (int i = 0; i < con->children->length; ++i) {
```

Comment:
Could you change the logic in this function to be:

* Iterate all children and render the tab/title
** Inside the loop, if the child is focused or focus_inactive, store it in a variable `active_child`
* Outside of the title loop, render the child stored in `active_child`

You then wouldn't need `get_title_textures` at all.

Reply:
Do you mean to render the tab/title for every child within the loop? I thought @SirCmpwn didn't want to take that approach and wanted the tab/title to be only rendered for one child and for the logic for the other tabs to come from there. Please correct me if I'm misunderstanding either one of you.

Reply:
> Do you mean to render the tab/title for every child within the loop?

Yes.

> I thought @SirCmpwn didn't want to take that approach and wanted the tab/title to be only rendered for one child and for the logic for the other tabs to come from there. Please correct me if I'm misunderstanding either one of you.

Yeah, I think there's a misunderstanding. The tab/title must be rendered for all children (which it is), but he's saying that the border width should be the same, regardless of whether the children use different widths or which child is focused. Cycling through each child should make the layout remain the same (ie. one border width used for all). I think the easiest and best option is to use the default border width stored in `config->border_thickness` rather than look for one in a child.

Reply:
Here's a quick example... this won't compile, but it should clarify the logic I would use:

render_container_tabbed():

	struct sway_container *active_child = NULL;
	for (int i = 0; i < con->children->length; ++i) {
		struct sway_container *child = con->children->items[i];
		if (child is focused) {
			class = &config->border_colors.focused;
			texture = &child->title_focused;
			active_child = child;
		} else if (child is focused_inactive) {
			class = &config->border_colors.focused_inactive;
			texture = &child->title_focused_inactive;
			active_child = child;
		} else if (child is unfocused) {
			class = &config->border_colors.unfocused;
			texture = &child->title_unfocused;
		}

        // Render title background for this tab here
        // Render single pixel lines above and below title here

		struct wlr_box box = {
			.x = con->x + config->border_thickness,
			.y = con->y + config->border_thickness,
			.width = con->width / con->children->length - config->border_thickness * 2,
			.height = con->title_height,
		};
		wlr_renderer_scissor(renderer, &box);
		wlr_render_texture(texture)
		wlr_renderer_scissor(renderer, NULL);
	}
	if (!sway_assert(active_child, "tabbed layout has no active child")) {
		return;
	}

	// Render left, bottom and right borders here - copy from another function and discard indicator border
	// Render actual view

Reply:
I think we should probably restructure the top-level loop to live entirely in the output code, instead of relying on the filtered map kind of deal. Then when we encounter a tabbed layout we can render its borders right there and worry about the children in recursive iterations.

Reply:
To accomplish this containers need to grow their own borders and some sort of bool should exist in the recursive function to indicate that child borders should no longer be drawn for subsequent iterations.

Reply:
Sorry, I have no clue what I'm talking about, disregard that

Reply:
Ah, that makes sense @RyanDwyer. Thanks for the explanation.